src/forcing/ForcingsEngineDataProvider.cpp

```diff
@@ -0,0 +1,428 @@
+    std::vector<int>                             var_divides_{};
+
+    /**
+     * Flat value cache vector.
+     * 
+     * Values are stored indexed on (timestep, divide_id, variable),
+     * such that the structure can be visualized as:
+     *
+     *   Timestep  : || T0       |          || T1  ...
+     *   Divide ID : || D0       | D1       || D0  ...
+     *   Variable  : || V1 V2 V3 | V1 V2 V3 || V1  ...
+     *   Value     : ||  9 11 31 |  3  4  5 || 10  ...
+     *
+     * Some notes for future reference:
+     * - Time complexity to update is approximately O(T*V*D),
+     *   where T = T2 - T1 is the number of timesteps between
+     *   two time points, V is the number of variables
+     *   and D is the number of divides. In general, D will dominate;
+     *   T will usually be 1, if updated at each time event; and, V will be
+     *   some small constant amount. Meaning, we should have about O(V*D).
+     */
+    ngen::mdarray<double> var_cache_{};
+};
+
+ForcingsEngine::ForcingsEngine(const std::string& init, size_type time_start, size_type time_end)
+  : time_start_(std::chrono::seconds{time_start})
+  , time_end_(std::chrono::seconds{time_end})
+{
+    check_runtime_requirements();
+
+    bmi_ = std::make_unique<models::bmi::Bmi_Py_Adapter>(
+        "ForcingsEngine",
+        init,
+        "NextGen_Forcings_Engine.BMIForcingsEngine",
+        true,
+        true,
+        utils::getStdOut()
+    );
+    
+    time_step_ = std::chrono::seconds{static_cast<int64_t>(bmi_->GetTimeStep())};
+    var_outputs_ = bmi_->GetOutputVarNames();
+    var_outputs_.erase(var_outputs_.begin()); // Erase "CAT-ID" from outputs
+    const auto time_dim = static_cast<size_type>((time_end_ - time_start_) / time_step_);
+    const auto id_dim   = static_cast<size_type>(bmi_->GetVarNbytes("CAT-ID") / bmi_->GetVarItemsize("CAT-ID"));
+    const auto var_dim  = var_outputs_.size();
+
+    bmi_->Update();
+    time_current_index_++;
+    
+    const auto* ptr = static_cast<int*>(bmi_->GetValuePtr("CAT-ID"));
+    var_divides_ = std::vector<int>(ptr, ptr + id_dim);
+    var_cache_ = decltype(var_cache_){{ time_dim, id_dim, var_dim }};
+
+    // Cache initial iteration
+    const auto size = var_divides_.size();
+    for (size_type vi = 0; vi < var_outputs_.size(); vi++) {
+        // Get current values for each variable
+        const auto var_span = boost::span<const double>{
+            static_cast<double*>(bmi_->GetValuePtr(var_outputs_[vi])),
+            size
+        };
+
+        for (size_type di = 0; di < size; di++) {
+            // Set current variable value for each divide
+            var_cache_.at({{time_current_index_, di, vi}}) = var_span[di];
+        }
+    }
+
+    time_current_index_++;
+};
+
+ForcingsEngine& ForcingsEngine::instance(
+    const std::string& init,
+    std::size_t time_start,
+    std::size_t time_end
+)
+{
+    static ForcingsEngine inst_{init, time_start, time_end};
```

Comment:
At least make this a `class` `static` variable, and not a function static. The singleton-ish nature is a little bit disrupted because of how the function arguments of the first call get used in construction.